lib/handler/configurator/headers_util.c

```diff
@@ -104,35 +124,36 @@ static int parse_header_node(h2o_configurator_command_t *cmd, yoml_t **node, yom
 static int on_config_header_2arg(h2o_configurator_command_t *cmd, h2o_configurator_context_t *ctx, int cmd_id, yoml_t *node,
                                  h2o_headers_command_t **headers_cmds)
 {
-    h2o_iovec_t *name, value;
     yoml_t **headers;
     size_t num_headers;
     h2o_headers_command_when_t when;
 
     if (parse_header_node(cmd, &node, &headers, &num_headers, &when) != 0)
         return -1;
 
+    h2o_iovec_t *name[num_headers], value[num_headers];
+    yoml_t *header[num_headers];
     int i;
     for (i = 0; i != num_headers; ++i) {
-        yoml_t *header = headers[i];
-        if (extract_name_value(header->data.scalar, &name, &value) != 0) {
-            h2o_configurator_errprintf(cmd, header, "failed to parse the value; should be in form of `name: value`");
+        header[i] = headers[i];
+        if (extract_name_value(header[i]->data.scalar, &name[i], &value[i]) != 0) {
+            h2o_configurator_errprintf(cmd, header[i], "failed to parse the value; should be in form of `name: value`");
             return -1;
         }
-        if (add_cmd(cmd, header, cmd_id, name, value, when, headers_cmds) != 0) {
-            if (!h2o_iovec_is_token(name))
-                free(name->base);
-            free(value.base);
-            return -1;
+    }
+    if (add_cmd(cmd, header, cmd_id, name, value, num_headers, when, headers_cmds) != 0) {
+        for (i = 0; i != num_headers; i++) {
+            if (!h2o_iovec_is_token(name[i]))
+                free(name[i]->base);
+            free(value[i].base);
         }
+        return -1;
     }
-
     return 0;
 }
 
-static int on_config_header_unset(h2o_configurator_command_t *cmd, h2o_configurator_context_t *ctx, yoml_t *node)
+static int on_config_unset_core(h2o_configurator_command_t *cmd, h2o_configurator_context_t *ctx, yoml_t *node, int type, int is_header)
```

Comment:
Would you mind changing `type` to `cmd_id`? I think that's the term we use.

I also tend to wonder if we could omit passing `is_header` as an argument, and rather determine how to parse the entries depending on the value of `cmd_id`.